### PR TITLE
WIP, discussion: Organizing tutorial resources and outputs

### PR DESCRIPTION
**Pull request by rossbar**
**tl;dr** - propose adding a `data/` and `_generated/` directory for files needed/produced by tutorials, respectively.

I wanted to start a discussion on what the policy will be for tutorial inputs and outputs. Most tutorials will have associated external files such as data or static images and often create output that needs to be saved to disk (images, files, etc). I think it would be good to formulate a policy on how these "external" resources will be handled in the tutorials. There are several things to consider:

## 1. Organization

Currently, all inputs are stored in the `content/` directory along with their corresponding tutorial. This means that the content directory will likely get pretty cluttered pretty quickly. Similarly, data files generated by the tutorials are currently being saved to disk in the `content/` directory, and are being tracked by version control.

## 2. Supporting all outputs (website, binder, and standalone notebooks)

This constraint is pretty tricky. The website and binder targets are straightforward because they both are generated from the entire repository --- all the necessary external files will be present and the paths correct. However, if a user downloads the standalone .ipynb, it won't be runnable if it depends on external files.

This PR addresses point 1 by adding two extra directories for external resources: one called `data/` for any data files that the tutorials depend on, and `_generated/` for any outputs (`.npz`s, `.png`s, `.csv`s, etc) generated by the tutorials. Organizing things this way helps reduce the clutter in `content/`, removes the possibility of accidently executing a file during site building that was not intended, and removes outputs from tracking in git. In the short term, I think an approach like this would help keep things organized for both authors and users.

It doesn't do anything to solve point 2 for standalone notebooks though, which I think is a more general problem. The way that I've dealt with this in the past is to have a policy that *all* supporting data/resources are to be hosted externally, and the tutorial has a bit of boilerplate to check for and download any missing resources, e.g.

```
if not os.path.exists('path/to/necessary/data'):
    import requests
    requests.get(url_hosting_necessary_data, **kwargs)
```

Of course, another possible solution would be to simply discourage downloading the standalone notebooks and recommend users to clone the repository and work from inside it as well. Just wanted to get some thoughts on how to handle this problem.

### CODE REVIEW

**Comment by melissawm** [comment]
Thanks, @rossbar ! That makes perfect sense. 

Your points all look reasonable to me. As for the external data resources when the user downloads an ipynb, I may be suggesting something too naive, but would it be possible to set the commands on the notebook to download the file from this github repo when the notebook is executed? This way we could guarantee reproducibility and reliability in case the external data changes- like my MA tutorial :)

**Comment by melissawm** [comment]
Hi @rossbar just wondering if you'd like to finish this or maybe do the same work but in a fresh PR, since so much of the rest has changed? If you don't have the time I can pick it up, no problem :) Cheers!

**Comment by rossbar** [comment]
Thanks for the ping! Yeah I would still like to clean this up a bit, but we've added so much new material since I opened this that it'd almost certainly be easier to start fresh :)
